### ISSUE TITLE
Unpredictable obsolete http blocking in Maven 3.8.1

### ISSUE THREAD

**Issue by quincy** [actions/virtual-environments#3362]
**Description**  
Maven was recently upgraded on the Linux runners to version 3.8.1.  This upgrade immediately broke builds for anyone using a maven repository with an insecure http URL.  

**Area for Triage**:  
Java

**Question, Bug, or Feature?**:  
Bug

**Virtual environments affected**  
- [ ] Ubuntu 16.04
- [x] Ubuntu 18.04
- [x] Ubuntu 20.04
- [ ] macOS 10.15
- [ ] macOS 11.0
- [ ] Windows Server 2016 R2
- [ ] Windows Server 2019

I've posted about this in the support community, but I'm copying that here as well.
https://github.community/t/unpredictable-obsolete-http-blocking-in-maven-3-8-1-on-github-actions/179252

I've updated the configuration in our `settings.xml` to unblock access to our self-hosted repository, and this has allowed builds to continue.

```
      - name: Setup Repo credentials
        # maven-settings-action v2.4.0
        uses: s4u/maven-settings-action@e967605742bd38de606cf6edb7c89349ce917c6a
        with:
          servers: |
            [{
                "id": "repo.releases.mirror",
                "username": "${{ secrets.REPO_USERNAME }}",
                "password": "${{ secrets.REPO_PASSWORD }}"
            }]
          mirrors: |
            [{
                "id": "repo.releases.mirror",
                "name": "repo.releases",
                "mirrorOf": "repo.releases",
                "url": "http://our-repo-url",
                "blocked": false
            }]
```

However, some branches have been failing during the deploy goal.

```
Error:  Failed to execute goal org.apache.maven.plugins:maven-deploy-plugin:3.0.0-M1:deploy (default-deploy) on project hats:
ArtifactDeployerException: Failed to retrieve remote metadata com.groupid:artifactid:version/maven-metadata.xml:
Could not transfer metadata com.groupid:artifactid:version/maven-metadata.xml from/to repo.snapshots (http://our-repo-url/):
authentication failed for http://our-repo-url/com/groupid/artifactd/version/maven-metadata.xml, status: 401 Unauthorized -> [Help 1]
```

These branches have no issue downloading dependencies from our repository using the unblocked http URL.  But when they get to the deploy step, and try to download the `maven-metadata.xml` from the same internal repository, it fails with a 401 Unauthorized.  Some branches don't exhibit this behavior at all.  If we recreate the branches that do have the issue, with the exact same commits, they build fine.  But then when we merge in changes from the main branch they start failing with the above issue again.

I suspect the issue is possibly in Github Actions itself, perhaps in a caching layer.  But So far I haven't been able to reliably recreate the issue on demand.  Some branches simply start failing because of this, and once that's happened we've been unable to get that branch to successfully deploy.

There is more detail in the community post itself.

https://github.community/t/unpredictable-obsolete-http-blocking-in-maven-3-8-1-on-github-actions/179252/2?u=quincy

The main takeaway is that I can make some changes and get the build to work, but then subsequent builds on the same branch start failing again, with a 401 Unauthorized to our repository.  Due to the nature of how it has been failing, I strongly suspect something more than misconfiguration on our end.  But I would be ecstatic if that's all it was.

I have seen the suggestions to downgrade to Maven 3.6.3 in #3327 and #3246 and will give that a try.  But I believe the inconsistent behavior I have seen still needs to be understood.

Thanks so much for your time.

**Comment by al-cheb** [comment]
Hi, @quincy 
We will take a look at this soon.

**Comment by al-cheb** [comment]
@quincy ,
Could you please reproduce the issue with?
`curl http://our-repo-url/com/groupid/artifactd/version/maven-metadata.xml -v -u user:pass`

**Comment by quincy** [comment]
Hi @al-cheb , here are the results with the sensitive bits redacted.

```
$ curl http://our-repo-url/com/groupid/artifactd/version/maven-metadata.xml -v -u username:password
*   Trying 1.1.1.1:8080...
* TCP_NODELAY set
* Connected to out-repo-url (54.205.76.146) port 8080 (#0)
* Server auth using Basic with user 'username'
> GET /repository/snapshots/com/appdetex/hats/HATS-2945-whois-seeding-newly-alerted-domains-SNAPSHOT/maven-metadata.xml HTTP/1.1
> Host: our-repo-url:8080
> Authorization: Basic xyz
> User-Agent: curl/7.68.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Date: Wed, 12 May 2021 12:15:04 GMT
< Set-Cookie: JSESSIONID=xyz;Path=/
< Expires: Thu, 01 Jan 1970 00:00:00 GMT
< Pragma: no-cache
< Cache-Control: no-cache
< Last-Modified: Wed, 12 May 2021 12:15:04 GMT
< Content-Length: 393
< Content-Type: application/xml
< Server: Jetty(8.1.14.v20131031)
< 
<?xml version="1.0" encoding="UTF-8"?>

<metadata>
  <groupId>groupid</groupId>
  <artifactId>artifactid</artifactId>
  <version>version</version>
  <versioning>
    <snapshot>
      <buildNumber>4</buildNumber>
      <timestamp>20210511.140919</timestamp>
    </snapshot>
    <lastUpdated>20210511140919</lastUpdated>
  </versioning>
</metadata>
* Connection #0 to host our-repo-url left intact
```

I tried this several times and always got the expected `maven-metadata.xml` in the response.

**Comment by al-cheb** [comment]
According this thread https://stackoverflow.com/questions/24830610/why-am-i-getting-a-401-unauthorized-error-in-maven, you can get auth 401 if a version already exists in the repository:

> You also get an error (and off the top of my head is also a 401) if you try to publish something to a releases repository and that version already exists in the repository. So you might find that by publishing from the command line it works, but then when you do it from a script it fails (because it didn't exist in the repository the first time around). Either publish using a different version number, or delete the old artefact on the server and republish.
> 
> The SNAPSHOTS repository (as opposed to the releases repository) allows you to overwrite a similarly numbered version, but your version number should have "-SNAPSHOT" at the end of it.

**Comment by quincy** [comment]
Hi @al-cheb , thank you for the suggestion.  The failure we are getting is during the deploy goal, but not when trying to upload the artifact.
```
Error:  Failed to execute goal org.apache.maven.plugins:maven-deploy-plugin:3.0.0-M1:deploy (default-deploy) on project hats:
ArtifactDeployerException: Failed to retrieve remote metadata com.groupid:artifactid:version/maven-metadata.xml:
Could not transfer metadata com.groupid:artifactid:version/maven-metadata.xml from/to repo.snapshots (http://our-repo-url/):
authentication failed for http://our-repo-url/com/groupid/artifactd/version/maven-metadata.xml, status: 401 Unauthorized -> [Help 1]
```

As you can see, this is happening while fetching the metadata.  So I believe publishing the same version shouldn't be an issue in this situation.  In any case, we do already use `-SNAPSHOT` on our PR builds.  `JIRA-1234-branch-name-SNAPSHOT` as an example.

**Comment by Darleev** [comment]
Hello @quincy,
Could you please try a solution provided in [the stackoverflow thread](https://stackoverflow.com/questions/65267872/github-action-cannot-download-artifact-from-github-package-repository)?(another [uses: s4u/maven-settings-action@v2.4.0](https://github.com/s4u/maven-settings-action) and remove cache).
We're looking forward to your reply.

**Comment by quincy** [comment]
> Hello @quincy,
> Could you please try a solution provided in [the stackoverflow thread](https://stackoverflow.com/questions/65267872/github-action-cannot-download-artifact-from-github-package-repository)?(another [uses: s4u/maven-settings-action@v2.4.0](https://github.com/s4u/maven-settings-action) and remove cache).
> We're looking forward to your reply.

Hi @Darleev, sorry for the delay.  I am trying to get back to looking into this for you.  Downgrading to maven 3.6 has fixed the issue for us for the time being.

**Comment by Darleev** [comment]
@quincy I guess, due to your update, it makes sense to create another issue to the Maven repository to clarify what was changed on their side.

**Comment by Darleev** [comment]
@quincy I'm going to close the issue until further is provided.
In case of any questions, feel free to contact us.

**Comment by Elikill58** [comment]
> Seeing the same issue here. We use `whelk-io/maven-settings-xml-action@v14` and the following for cache, which doesn't include `~/.m2/settings.xml`
> 
> ```
>       - name: Cache local Maven repository
>         uses: actions/cache@v2
>         with:
>           path: ~/.m2/repository
>           key: ${{ runner.os }}-maven-${{ hashFiles('**/pom.xml') }}
>           restore-keys: |
>             ${{ runner.os }}-maven-
> ```
> 

I'm trying to fix the same issue, and it doesn't seems to work :

![image](https://user-images.githubusercontent.com/17381066/132962947-b6705986-2234-4533-b43a-3f47ad9c61f6.png)

And there is cache, such as tell in few warnings :
`This failure was cached in the local repository and resolution will not be reattempted until the update interval of maven-default-http-blocker has elapsed or updates are forced.`

Also, for the issue, I tried the fix if [this SO answer](https://stackoverflow.com/a/67002852/10952503) but it's using `http://0.0.0.0/` URL (even if I well config it).

So someone know how I can fix it ?

**Comment by quincy** [comment]
After we downgraded to maven 3.6 we also put a load balancer to handle https in front of our maven repo.  I would recommend one of these two routes as they are both relatively easy and don't fight what Apache is trying to do with Maven.
